File: src/sync/desktop_data.py
"""有道云笔记桌面客户端本地数据读取。

提供两个可选能力：
1. 冷启动种子 — 首次运行时从桌面 DB 导入文件元数据到 sync_metadata.db
2. domain=0 文件本地读取 — 省一次 HTTP 下载
"""

import logging
import os
import platform
import sqlite3
from typing import Optional, Dict, Tuple

from src.common import FileId
from src.sync.scanner import map_cloud_name

logger = logging.getLogger(__name__)


def find_desktop_data_dir() -> Optional[str]:
    """定位桌面客户端的 ynote-data 目录。

    桌面客户端的数据目录结构：
      %APPDATA%/ynote-desktop/<user>/ynote-data/
    其中 <user> 是邮箱地址，通过遍历 ynote-desktop 下的子目录来发现。
    """
    system = platform.system()
    if system == "Windows":
        base = os.environ.get("APPDATA", "")
    elif system == "Darwin":
        base = os.path.expanduser("~/Library/Application Support")
    else:
        base = os.path.expanduser("~/.config")

    ynote_dir = os.path.join(base, "ynote-desktop")
    if not os.path.isdir(ynote_dir):
        return None

    for entry in os.listdir(ynote_dir):
        data_dir = os.path.join(ynote_dir, entry, "ynote-data")
        if os.path.isdir(data_dir):
            db_candidates = [f for f in os.listdir(data_dir)
                             if f.endswith(".db") and not f.endswith("-content.db")
                             and not f.endswith("-search.db")]
            if db_candidates:
                return data_dir
    return None


def find_desktop_db(data_dir: str) -> Optional[str]:
    """在 ynote-data 目录中找到主 SQLite 数据库文件。"""
    if not data_dir or not os.path.isdir(data_dir):
        return None
    for f in os.listdir(data_dir):
        if f.endswith(".db") and not f.endswith("-content.db") \
                and not f.endswith("-search.db"):
            path = os.path.join(data_dir, f)
            if os.path.isfile(path):
                return path
    return None


def _build_path_map(conn: sqlite3.Connection) -> Dict[str, str]:
    """从 note_book 表构建 folder_id → 路径 映射。

    note_book 表的 parentId 形成一棵树，根节点的 parentId 是空或指向自身。
    """
    cur = conn.cursor()
    cur.execute("SELECT name FROM sqlite_master WHERE type='table' AND name='note_book'")
    if not cur.fetchone():
        return {}

    cur.execute("""SELECT fileId, title, parentId FROM note_book
                   WHERE del = 0 OR del IS NULL""")
    folders = {}
    for fid, title, pid in cur.fetchall():
        folders[fid] = {"title": title or "", "parent_id": pid or ""}

    path_cache: Dict[str, str] = {}

    def resolve(fid: str, depth: int = 0) -> str:
        if fid in path_cache:
            return path_cache[fid]
        if depth > 50 or fid not in folders:
            return ""
        info = folders[fid]
        pid = info["parent_id"]
        if not pid or pid == fid:
            path_cache[fid] = ""
            return ""
        parent_path = resolve(pid, depth + 1)
        path = f"{parent_path}/{info['title']}" if parent_path else info["title"]
        path_cache[fid] = path
        return path

    for fid in folders:
        resolve(fid)

    return path_cache


def seed_metadata_from_desktop(metadata, data_dir: str = None) -> int:
    """从桌面客户端 DB 导入文件和目录元数据（冷启动种子）。

    :param metadata: SyncMetadata 实例
    :param data_dir: ynote-data 目录，为 None 时自动发现
    :return: 导入的条目数
    """
    if not data_dir:
        data_dir = find_desktop_data_dir()
    if not data_dir:
        logger.debug("桌面客户端数据目录未找到")
        return 0

    db_path = find_desktop_db(data_dir)
    if not db_path:
        logger.debug("桌面客户端数据库未找到")
        return 0

    try:
        conn = sqlite3.connect(f"file:{db_path}?mode=ro", uri=True)
        conn.row_factory = sqlite3.Row
    except sqlite3.Error as e:
        logger.warning(f"打开桌面数据库失败: {e}")
        return 0

    try:
        folder_paths = _build_path_map(conn)
        cur = conn.cursor()
        count = 0

        with metadata.batch():
            for fid, path in folder_paths.items():
                if path:
                    cur.execute(
                        "SELECT parentId FROM note_book WHERE fileId = ?", (fid,))
                    row = cur.fetchone()
                    parent_id = row["parentId"] if row else ""
                    metadata.set_dir_info(path, fid, parent_id)
                    count += 1

            cur.execute("""SELECT fileId, title, parentId, modifyTime,
                                  createTime, domain, version
                           FROM note WHERE del = 0 AND dir = 0""")
            for row in cur.fetchall():
                fid = row["fileId"]
                parent_id = row["parentId"] or ""
                folder_path = folder_paths.get(parent_id, "")
                title = row["title"] or ""

                local_name = map_cloud_name(title)
                rel_path = f"{folder_path}/{local_name}" if folder_path else local_name

                mtime = row["modifyTime"] or 0
                if mtime > 1_000_000_000_000:
                    mtime = mtime // 1000

                ctime = row["createTime"] or 0
                if ctime > 1_000_000_000_000:
                    ctime = ctime // 1000

                metadata.set_file_info(
                    local_path=rel_path,
                    file_id=fid,
                    cloud_mtime=mtime,
                    parent_id=parent_id,
                    domain=row["domain"],
                    create_time=ctime,
                )
                count += 1

            if count > 0:
                max_version = 0
                cur.execute("SELECT MAX(version) FROM note WHERE del = 0")
                r = cur.fetchone()
                if r and r[0]:
                    max_version = r[0]
                metadata.set_state("last_cloud_version", str(max_version))

        metadata.save()
        logger.info(f"从桌面客户端导入 {count} 条元数据 (种子)")
        return count

    except sqlite3.Error as e:
        logger.warning(f"读取桌面数据库失败: {e}")
        return 0
    finally:
        conn.close()


def read_desktop_file(file_id: FileId, data_dir: str = None) -> Optional[bytes]:
    """从桌面客户端本地缓存读取 domain=0 文件内容。

    桌面客户端将 domain=0 文件的 XML 存储在 file/<bucket>/<fileId>。
    bucket 是 fileId 首字符的小写。

    :return: 文件字节内容，未找到时返回 None
    """
    if not file_id:
        return None

    if not data_dir:
        data_dir = find_desktop_data_dir()
    if not data_dir:
        return None

    bucket = file_id[0].lower()
    path = os.path.join(data_dir, "file", bucket, file_id)
    if not os.path.isfile(path):
        return None

    try:
        with open(path, "rb") as f:
            content = f.read()
        if content:
            logger.debug(f"从桌面缓存读取: {file_id} ({len(content)} bytes)")
        return content
    except OSError as e:
        logger.debug(f"读取桌面缓存文件失败: {file_id} - {e}")
        return None

File: ts-src/src/desktop-data.ts
/**
 * Read metadata from the Youdao Note desktop client's local SQLite DB.
 *
 * Two capabilities:
 * 1. Cold-start seed: first run imports file metadata into sync_metadata.db
 * 2. domain=0 file local read (saves an HTTP download)
 *
 * Matches Python desktop_data.py.
 */
import { existsSync, readdirSync, readFileSync } from 'node:fs';
import { join, basename } from 'node:path';
import { platform, env } from 'node:process';
import { homedir } from 'node:os';
import Database from 'better-sqlite3';
import type { MetadataStore } from './metadata/store.js';
import type { DirId, FileId } from './types/common.js';
import { mapCloudName } from './scan/name.js';

export function findDesktopDataDir(): string | null {
  let base: string;
  if (platform === 'win32') {
    base = env['APPDATA'] ?? '';
  } else if (platform === 'darwin') {
    base = join(homedir(), 'Library', 'Application Support');
  } else {
    base = join(homedir(), '.config');
  }

  const ynoteDir = join(base, 'ynote-desktop');
  if (!existsSync(ynoteDir)) return null;

  for (const entry of readdirSync(ynoteDir)) {
    const dataDir = join(ynoteDir, entry, 'ynote-data');
    if (!existsSync(dataDir)) continue;
    const dbCandidates = readdirSync(dataDir).filter(
      (f) => f.endsWith('.db') && !f.endsWith('-content.db') && !f.endsWith('-search.db'),
    );
    if (dbCandidates.length > 0) return dataDir;
  }

  return null;
}

export function findDesktopDb(dataDir: string): string | null {
  if (!dataDir || !existsSync(dataDir)) return null;
  for (const f of readdirSync(dataDir)) {
    if (f.endsWith('.db') && !f.endsWith('-content.db') && !f.endsWith('-search.db')) {
      const path = join(dataDir, f);
      if (existsSync(path)) return path;
    }
  }
  return null;
}

function buildPathMap(conn: Database.Database): Map<string, string> {
  const hasTable = conn.prepare(
    "SELECT name FROM sqlite_master WHERE type='table' AND name='note_book'",
  ).get();
  if (!hasTable) return new Map();

  const rows = conn.prepare(
    "SELECT fileId, title, parentId FROM note_book WHERE del = 0 OR del IS NULL",
  ).all() as Array<{ fileId: string; title: string; parentId: string }>;

  const folders = new Map<string, { title: string; parentId: string }>();
  for (const row of rows) {
    folders.set(row.fileId, { title: row.title || '', parentId: row.parentId || '' });
  }

  const cache = new Map<string, string>();

  function resolve(fid: string, depth: number): string {
    if (cache.has(fid)) return cache.get(fid)!;
    if (depth > 50 || !folders.has(fid)) return '';
    const info = folders.get(fid)!;
    const pid = info.parentId;
    if (!pid || pid === fid) {
      cache.set(fid, '');
      return '';
    }
    const parentPath = resolve(pid, depth + 1);
    const path = parentPath ? `${parentPath}/${info.title}` : info.title;
    cache.set(fid, path);
    return path;
  }

  for (const fid of folders.keys()) {
    resolve(fid, 0);
  }

  return cache;
}

/**
 * Seed sync metadata from the desktop client's local DB (cold-start).
 * Returns the number of imported entries.
 */
export function seedMetadataFromDesktop(meta: MetadataStore, dataDir?: string): number {
  const dir = dataDir ?? findDesktopDataDir();
  if (!dir) return 0;

  const dbPath = findDesktopDb(dir);
  if (!dbPath) return 0;

  let conn: Database.Database;
  try {
    conn = new Database(dbPath, { readonly: true, fileMustExist: true });
  } catch {
    return 0;
  }

  try {
    const folderPaths = buildPathMap(conn);
    let count = 0;

    meta.batch(() => {
      for (const [fid, path] of folderPaths) {
        if (!path) continue;
        const row = conn.prepare(
          'SELECT parentId FROM note_book WHERE fileId = ?',
        ).get(fid) as { parentId: string } | undefined;
        const parentId = row?.parentId ?? '';
        meta.setDirInfo(path, fid as DirId, parentId as DirId);
        count++;
      }

      const noteRows = conn.prepare(
        "SELECT fileId, title, parentId, modifyTime, createTime, domain, version FROM note WHERE del = 0 AND dir = 0",
      ).all() as Array<{
        fileId: string; title: string; parentId: string;
        modifyTime: number; createTime: number; domain: number; version: number;
      }>;

      for (const row of noteRows) {
        const parentId = row.parentId || '';
        const folderPath = folderPaths.get(parentId) ?? '';
        const title = row.title || '';
        const localName = mapCloudName(title);
        const relPath = folderPath ? `${folderPath}/${localName}` : localName;

        let mtime = row.modifyTime || 0;
        if (mtime > 1_000_000_000_000) mtime = Math.floor(mtime / 1000);

        let ctime = row.createTime || 0;
        if (ctime > 1_000_000_000_000) ctime = Math.floor(ctime / 1000);

        meta.setFileInfo(relPath, {
          fileId: row.fileId as FileId,
          cloudMtime: mtime,
          localMtime: 0,
          parentId: parentId as DirId,
          domain: row.domain,
          createTime: ctime,
        });
        count++;
      }

      if (count > 0) {
        let maxVersion = 0;
        const vRow = conn.prepare(
          'SELECT MAX(version) as mv FROM note WHERE del = 0',
        ).get() as { mv: number | null } | undefined;
        if (vRow?.mv) maxVersion = vRow.mv;
        meta.setState('last_cloud_version', String(maxVersion));
      }
    });

    meta.save();
    return count;
  } catch {
    return 0;
  } finally {
    conn.close();
  }
}

/**
 * Read a domain=0 file from the desktop client's local cache.
 * The desktop client stores XML in file/<bucket>/<fileId>.
 */
export function readDesktopFile(fileId: FileId, dataDir?: string): Buffer | null {
  if (!fileId) return null;
  const dir = dataDir ?? findDesktopDataDir();
  if (!dir) return null;

  const bucket = fileId[0]!.toLowerCase();
  const path = join(dir, 'file', bucket, fileId);
  if (!existsSync(path)) return null;

  try {
    return readFileSync(path);
  } catch {
    return null;
  }
}
